File: util/M1/Final/src/com/jackokie/objects/Position.java
package com.jackokie.objects;

public class Position implements Comparable<Position>{
	private double longitude = 0;
	private double latitude = 0;

	public Position() {

	}

	public Position(double shopLong, double shopLat) {
		this.longitude = shopLong;
		this.latitude = shopLat;
	}

	public double getLongitude() {
		return longitude;
	}

	public void setLongitude(double longitude) {
		this.longitude = longitude;
	}

	public double getLatitude() {
		return latitude;
	}

	public void setLatitude(double latitude) {
		this.latitude = latitude;
	}

	@Override
	public boolean equals(Object pos) {
		if (pos instanceof Position) {
			Position position = (Position) pos;
			if ((Math.abs(position.getLatitude() - this.latitude) < 0.000000001)
					&& Math.abs(position.getLongitude() - this.longitude) < 0.000000001) {
				return true;
			}
		}
		return false;
	}

	@Override
	public int hashCode() {
		return (String.valueOf(longitude) + String.valueOf(latitude)).hashCode();
	}

	public boolean near(Position posNew, double DIF_LONG_LAT) {
		// 表示两个位置是否相近
		double difLong = Math.abs(this.longitude - posNew.getLongitude());
		double difLat = Math.abs(this.latitude - posNew.getLatitude());
		if (difLong < DIF_LONG_LAT && difLat < DIF_LONG_LAT) {
			return true;
		} else {
			return false;
		}
	}

	// 判断是否在店铺的作用域
	public boolean inShopZone(Position shopPos) {
		return (shopPos.getLongitude() - longitude > 0) && (shopPos.getLatitude() - latitude > 0);
	}

	public boolean upNear(Position pos, double DEF) {
		double difLong = pos.getLongitude() - this.longitude;
		double difLat = pos.getLatitude() - this.latitude;
		if (difLong < DEF && difLat < DEF) {
			return true;
		} else {
			return false;
		}
	}

	public double getDis(Position pos) {
		double longTemp = pos.getLongitude();
		double latTemp = pos.getLatitude();
		double difLong = Math.abs(longitude - longTemp);
		double difLat = Math.abs(latitude - latTemp);
		double b = (latitude + latTemp) / 2.0; // 平均纬度
		double Lx = degree2Rad(difLong) * 6367000.0 * Math.cos(degree2Rad(b)); // 东西距离
		double Ly = 6367000.0 * degree2Rad(difLat); // 南北距离
		double dis = Math.sqrt(Lx * Lx + Ly * Ly); // 用平面的矩形对角距离公式计算总距离
		return dis;
	}

	private double degree2Rad(Double degree) {
		return degree * Math.PI / 180.0;
	}

	@Override
	public int compareTo(Position pos) {
		Double posLong = pos.getLongitude();
		Double posLat = pos.getLatitude();
		Double longO = longitude;
		Double latO = latitude;
		if (longO.compareTo(posLong) != 0) {
			return longO.compareTo(posLong);
		} else {
			return latO.compareTo(posLat);
		}
	}

	@Override
	public String toString() {
		return "(" + longitude + "," + latitude + ")";
	}
}

File: util/M1/ChinaUnicomFS/objects/Position.py
#!/usr/bin/env python
# encoding: utf-8


"""
@version: v1.0
@author: jackokie
@license: Jackokie Licence 
@contact: jackokie@qq.com
@site: http://www.jackokie.com
@software: PyCharm
@file: Position.py
@time: 2016/12/8 20:12
"""

import math


class Position():
	def __init__(self, long=0, lat=0):
		self._longitude = long
		self._latitude = lat
	
	def get_longitude(self):
		return self._longitude
	
	def set_longitude(self, long):
		self._longitude = long
	
	def get_latitude(self):
		return self._latitude
	
	def set_latitude(self, lat):
		self._latitude = lat
	
	def equal(self, pos):
		if isinstance(pos, Position):
			if (abs(pos.get_latitude() - self._latitude) < 0.0000001) & (
						abs(pos.get_longitude() - self._longitude) < 0.0000001):
				return True
		return False
	
	def near(self, pos, DIF_LONG_LAT):
		'''两个位置是否相近'''
		dif_long = abs(pos.get_longitude() - self._longitude)
		dif_lat = abs(pos.get_latitude() - self._latitude)
		
		if (dif_long < DIF_LONG_LAT) & (dif_lat < DIF_LONG_LAT):
			return True
		else:
			return False
	
	def upNear(self, pos, DEF):
		dif_long = pos.get_longitude() - self._longitude
		dif_lat = pos.get_latitude() - self._latitude
		if dif_long < DEF & dif_lat < DEF:
			return True
		else:
			return False
	
	def inEffectZone(self, pos):
		'''判断是否在店铺的作用域以内'''
		return (pos.get_longitude() > self._longitude) & (pos.get_latitude() > self._latitude)
	
	def degree2rad(self, degree):
		return degree * math.pi / 180.0
	
	def get_dis(self, pos):
		dif_long = abs(self._longitude - pos.get_longitude())
		dif_lat = abs(self._latitude - pos.get_latitude())
		b = (self._latitude + pos.get_latitude() / 2.0)
		dx = self.degree2rad(dif_long) * 6367000.0 * math.cos(self.degree2rad(b))
		dy = 6367000.0 * self.degree2rad(dif_lat)
		dis = math.sqrt(dx * dx + dy * dy)
		return dis
	
	def __eq__(self, other):
		if not isinstance(other, Position):
			raise (TypeError, "can't cmp other type to Post!")
		if self._longitude == other.get_longitude() and self._latitude == other.get_latitude():
			return True
		else:
			return False
	def __hash__(self):
		return hash(str(self._longitude) + str(self._latitude))